rhine-terminal/TerminalSimple.hs

```diff
@@ -0,0 +1,100 @@
+{- | Example application for the @rhine-terminal@ library. -}
+
+{-# LANGUAGE DataKinds #-}
+{-# LANGUAGE FlexibleInstances #-}
+{-# LANGUAGE FlexibleContexts #-}
+{-# LANGUAGE LambdaCase #-}
+{-# LANGUAGE MultiParamTypeClasses #-}
+{-# LANGUAGE OverloadedStrings #-}
+{-# LANGUAGE TypeFamilies #-}
+{-# LANGUAGE UndecidableInstances #-}
+{-# LANGUAGE ScopedTypeVariables #-}
+
+import Prelude hiding (putChar)
+
+-- rhine
+import Control.Monad.Schedule ()
+
+import Data.Text (Text)
+import qualified Data.Text as T
+import System.Terminal
+import System.Terminal.Internal
+import FRP.Rhine
+import System.IO hiding (putChar)
+
+import FRP.Rhine.Terminal (TerminalEventClock(..))
+import System.Exit (exitSuccess)
+
+-- Clocks
+
+data Input = Char Char
+           | Space | Backspace | Enter
+           | Exit
+
+type InputClock = SelectClock (TerminalEventClock LocalTerminal) Input
+
+inputClock :: LocalTerminal -> InputClock
+inputClock term = SelectClock { mainClock = TerminalEventClock term, select = selectKey }
+  where
+    selectKey :: Tag (TerminalEventClock LocalTerminal) -> Maybe Input
+    selectKey = \case
```

Comment:
Why not pattern match directly in the function?

Reply:
It makes it a bit more verbose with extra parens and duplication of  `select` but I still made the change:
```haskell
select (Right (KeyEvent (CharKey k) m))  = Just (Char k m)
select (Right (KeyEvent SpaceKey _))     = Just Space
select (Right (KeyEvent BackspaceKey _)) = Just Backspace
select (Right (KeyEvent EnterKey _))     = Just Enter
select (Left _)                          = Just Exit
select _                                 = Nothing
```

vs 

```haskell
select = \case
  Right (KeyEvent (CharKey k) _)  -> Just (Char k)
  Right (KeyEvent SpaceKey _)     -> Just Space
  Right (KeyEvent BackspaceKey _) -> Just Backspace
  Right (KeyEvent EnterKey _)     -> Just Enter
  Left _                          -> Just Exit
  _                               -> Nothing
```

Reply:
True. I'm starting to like your style here :)